File: 46-permutations/46-permutations.cpp
class Solution {
public:
    vector<vector<int>>ans;
    
    void permute1(vector<vector<int>> &ans,int idx , vector<int>& nums){
        if(idx==nums.size()){
            ans.push_back(nums);
        }
        for(int i=idx;i<nums.size();i++){
            swap(nums[i],nums[idx]);
            permute1(ans,idx+1,nums);
            swap(nums[i],nums[idx]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        permute1(ans,0,nums);
        return ans;
    }
};

File: 46-permutations/46-permutations.java
class Solution {
    public int permute1(int idx,int[] nums,List<List<Integer>> ans,List<Integer> sAns) {
        if(idx==nums.length){
            ArrayList<Integer> base = new ArrayList<>(sAns);
            ans.add(base);
            return 1;
        }
      int  count=0;
        for(int i=0;i<nums.length;i++){
            if(nums[i]>-11){
                int val = nums[i];
                nums[i]=-11;
                sAns.add(val);

              count+=  permute1(idx+1,nums,ans,sAns);

                nums[i]=val;
                sAns.remove(sAns.size()-1);
            }
        }
        return count;
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> sAns = new ArrayList<>();
        
         permute1(0,nums,ans,sAns);
        return ans;
    }
}